src/rustllvm/PassWrapper.cpp

```diff
@@ -503,8 +504,133 @@ LLVMRustWriteOutputFile(LLVMTargetMachineRef Target, LLVMPassManagerRef PMR,
   return LLVMRustResult::Success;
 }
 
+
+// Callback to demangle function name
+// Parameters:
+// * name to be demangled
+// * name len
+// * output buffer
+// * output buffer len
+// Returns len of demangled string, or 0 if demangle failed.
+typedef size_t (*DemangleFn)(const char*, size_t, char*, size_t);
+
+
+namespace {
+
+class RustAssemblyAnnotationWriter : public AssemblyAnnotationWriter {
+  DemangleFn Demangle;
+  static const size_t BufCap = 500;
+  char Buf[BufCap];
```

Comment:
Could the output buffer size derived from the input symbol instead? Something like a std::vector that only ever expands when it sees an input symbol longer than its current output buffer.